### ISSUE TITLE
add_roles seems to fail sem-randomly

### ISSUE THREAD

**Issue by Lazyboy007** [Rapptz/discord.py#1726]
Trying to define a command allowing users to follow courses in a discord I'm moderating. The bot has two commands: follow_course and unfollow_course, that give or remove a list of roles to the user requesting them. Here's the code for follow_course (unfollow is pretty much the same). The prints are used to debug:

```python
    @client.command(aliases=['follow', 'fc'], pass_context=True)
    async def follow_course(context):
        """Allows an user to follow a list of courses."""
        m = context.message
        u = m.author

        message = m.content
        if message.find(" ") > 0:
            success = ""
            fail = ""
            refused = ""
            print(u)
            for name in message.split(" ")[1:]:
                role = discord.utils.get(m.server.roles, name=name.upper())
                print(role)
                if role:
                    annonceur = discord.utils.get(m.server.roles, name="Annonceur")
                    print(annonceur)
                    if role >= annonceur or role in u.roles:
                        print("illegal")
                        refused += name + " "
                    else:
                        await client.add_roles(u, role)
                        print("success")
                        success += name + " "
                else:
                    print("rip")
                    fail += name + " "
            full = "You successfully followed: " + success.strip() + "\n" if success else ""
            full += "Couldn't follow: " + refused.strip() + "\n" if refused else ""
            full += "Couldn't find: " + fail.strip() if fail else ""
            print(full)
            print()
            await client.send_message(u, full.strip())
        else:
            await client.say("Please provide a course to follow")
```

What i am noticing, is that add_roles seems to fail. Here is a sample print from one request I made with 4 courses where one failed:

> lazyboy007#0007
> LBIR1315
> Annonceur
> success
> LINGI2142
> Annonceur
> success
> LINGI1123
> Annonceur
> success
> LINGI1131
> Annonceur
> success
> You successfully followed: LBIR1315 LINGI2142 LINGI1123 LINGI1131

Here, LINGI1123 failed. A friend of mine tried the same command, got the same output (except for the user tag obviously) and in his case, LBIR1315 failed as well. No matter how many times we repeat this, the same happens. I hope it is not a mistake on my part in using the add_roles command, but it looks to be something in that command that is not working consistently.

Thank you for looking into it.

**Comment by Gelbpunkt** [comment]
If you're using ext.commands, what you obviously are, you can do
`async def follow_course(context, *roles: discord.Role):`
roles will be a list of role objects and makes your code cleaner.

**Comment by Lazyboy007** [comment]
Ok, seems that putting them all in a list and then giving all roles to the user using a single add_roles call works. Still strange that consecutive calls to add_roles on the same user ends up not working.

**Comment by Vexs** [comment]
This is a caching bug- detailed in this issue: #56

Tldr, don't call add/remove roles quickly. Take advantage of their plurality, and if you have to add and remove at the same time, use replace roles.

**Comment by Lazyboy007** [comment]
The bug is reported as fixed on that specific issue, but yeah for me it's working using the list. I'll leave it up to Rapptz to decide if they want to close this.

**Comment by Phxntxm** [comment]
You're misunderstanding what happened in that situation. That was fixed for the *rewrite* branch, not for async. The async branch is under a freeze as it will be deprecated and replaced with the rewrite....the solution here is to either switch to rewrite, or to use replace_roles....this won't be fixed in the async branch.

**Comment by Lazyboy007** [comment]
Ah I see thanks for clarifying
